lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp

```diff
@@ -232,6 +232,49 @@ struct ConvertLayoutOpConversion
       return success();
     }
 
+    if (isMmaToMmaShortcut(srcTy, dstTy)) {
```

Comment:
we used to have lowering for this but it went away in https://github.com/triton-lang/triton/pull/4125. @jlebar Do you expect this case should be handled? If not I suggest we revive the pattern
```
struct ConvertLayoutOpOptimizedConversion
    : public ConvertOpToLLVMPattern<triton::gpu::ConvertLayoutOp> {
public:
  using ConvertOpToLLVMPattern<
      triton::gpu::ConvertLayoutOp>::ConvertOpToLLVMPattern;
  LogicalResult
  matchAndRewrite(triton::gpu::ConvertLayoutOp op, OpAdaptor adaptor,
                  ConversionPatternRewriter &rewriter) const override {
    RankedTensorType srcTy = op.getSrc().getType();
    RankedTensorType dstTy = op.getType();
    Attribute srcLayout = srcTy.getEncoding();
    Attribute dstLayout = dstTy.getEncoding();
    // forwarding on mma->mma shortcut, lower distributed->distributed otherwise
    if (isa<NvidiaMmaEncodingAttr>(srcLayout) &&
        isa<NvidiaMmaEncodingAttr>(dstLayout)) {
      if (isMmaToMmaShortcut(srcTy, dstTy)) {
        return lowerMmaToMma(op, adaptor, rewriter);
      }
    }
    return failure();
  }
```

Reply:
> Do you expect this case should be handled?

IIRC the mma->mma shortcut was moving data between registers in the same thread.  https://github.com/triton-lang/triton/pull/4125 handles this using LLs.

I guess I'm not sure what the question is?  :)

Reply:
so you are saying it should be handled by the convert layout of register to register. @htyu can you debug why this is not the case?

Reply:
I see. Looks like hitting a TODO in linear layout. I guess we want to fix that instead?

https://github.com/triton-lang/triton/blob/d0418917ed2bb86dc7a3cbcc7b86b075c5d1ee4d/lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp#L414

Reply:
weird, I thought this convert would only move data within a thread

Reply:
The checking for intra-thread move fails (https://github.com/triton-lang/triton/blob/d0418917ed2bb86dc7a3cbcc7b86b075c5d1ee4d/lib/Tools/LinearLayout.cpp#L682). I'm still understanding how it is supposed to work. But do you think linear layout can handle the case where number of input registers isn't the same as the output?

Reply:
Some more information:
```

[linear_layout]: this->divideRight(divisor)=candidate_quotient
[linear_layout]: this:
 - register=1 -> (1, 0, 0, 0)
   register=2 -> (2, 0, 0, 0)
   register=4 -> (4, 0, 0, 0)
   register=8 -> (8, 0, 0, 0)
   register=16 -> (16, 0, 0, 0)
   register=32 -> (0, 1, 0, 0)
   register=64 -> (32, 0, 0, 0)
 - lane=1 -> (0, 2, 0, 0)
   lane=2 -> (0, 0, 0, 0)
   lane=4 -> (0, 4, 0, 0)
   lane=8 -> (0, 8, 0, 0)
   lane=16 -> (0, 16, 0, 0)
 - warp=1 -> (0, 0, 1, 0)
   warp=2 -> (0, 0, 2, 0)
 - block is a size 1 dimension
where out dims are: [register (size 64), lane (size 32), warp (size 4), block (size 1)]


[linear_layout]: divisor:
 - lane=1 -> (1, 0, 0)
   lane=2 -> (2, 0, 0)
   lane=4 -> (4, 0, 0)
   lane=8 -> (8, 0, 0)
   lane=16 -> (16, 0, 0)
 - warp=1 -> (0, 1, 0)
   warp=2 -> (0, 2, 0)
 - block is a size 1 dimension
where out dims are: [lane (size 32), warp (size 4), block (size 1)]



[linear_layout]: newOutDims: register=64
[linear_layout]: checkInvariants: 
 - register=1 -> (1)
   register=2 -> (2)
   register=4 -> (4)
   register=8 -> (8)
   register=16 -> (16)
   register=32 -> (0)
   register=64 -> (32)
where out dims are: [register (size 64)]

[linear_layout]: candidate_quotient:
- register=1 -> (1)
   register=2 -> (2)
   register=4 -> (4)
   register=8 -> (8)
   register=16 -> (16)
   register=32 -> (0)
   register=64 -> (32)
where out dims are: [register (size 64)]
```